File: backend/account/latin_validation.py
"""Require Latin-script letters for names and address fields (Sendcloud / UK shipping)."""

from __future__ import annotations

import unicodedata

LATIN_SCRIPT_ERROR = "Use Latin characters only"


def is_latin_script_text(value: str | None) -> bool:
    """
    Return True when ``value`` is empty or every *letter* is Latin-script.

    Digits, whitespace, punctuation, and symbols are allowed (including
    typographic marks from mobile keyboards). Cyrillic and other non-Latin
    letters are rejected. Accented Latin (e.g. José, Müller) is OK.
    """
    text = value if value is not None else ""
    if not str(text).strip():
        return True

    for ch in str(text):
        # Only alphabetic characters are constrained to Latin script.
        # Punctuation / symbols (curly apostrophes, dashes, £, etc.) pass.
        if not ch.isalpha():
            continue
        name = unicodedata.name(ch, "")
        if not name.startswith("LATIN"):
            return False
    return True


def latin_script_error(value: str | None) -> str | None:
    """Return an error message when ``value`` is not Latin-script text."""
    if is_latin_script_text(value):
        return None
    return LATIN_SCRIPT_ERROR


def add_latin_script_errors(form, field_names: tuple[str, ...] | list[str]) -> None:
    """
    Attach Latin-script errors for any non-empty cleaned values on ``field_names``.

    Skips fields that are not on the form or already have errors.
    """
    cleaned = getattr(form, "cleaned_data", None) or {}
    for name in field_names:
        if name not in getattr(form, "fields", {}):
            continue
        if form.errors.get(name):
            continue
        message = latin_script_error(cleaned.get(name))
        if message:
            form.add_error(name, message)

File: frontend-marketplace/src/utils/latinValidation.ts
/**
 * Latin-script validation for names and address fields (matches backend).
 * Only letters are constrained to Latin script; punctuation and symbols
 * (including typographic marks from mobile keyboards) are allowed.
 * Rejects Cyrillic and other non-Latin letters; accented Latin (José) is allowed.
 */

export const LATIN_SCRIPT_ERROR = "Use Latin characters only";

export function isLatinScriptText(value: string | null | undefined): boolean {
  const text = value ?? "";
  if (!text.trim()) {
    return true;
  }

  for (const ch of text) {
    // Only alphabetic characters are constrained to Latin script.
    // Punctuation / symbols (curly apostrophes, dashes, £, etc.) pass.
    if (!/\p{L}/u.test(ch)) {
      continue;
    }
    if (!/\p{Script=Latin}/u.test(ch)) {
      return false;
    }
  }
  return true;
}

export function latinScriptError(
  value: string | null | undefined
): string | null {
  return isLatinScriptText(value) ? null : LATIN_SCRIPT_ERROR;
}

/** Add Latin-script errors for non-empty fields that are not already in `errors`. */
export function applyLatinScriptErrors(
  fields: Record<string, string | null | undefined>,
  errors: Record<string, string>
): void {
  for (const [key, value] of Object.entries(fields)) {
    if (errors[key]) {
      continue;
    }
    const message = latinScriptError(value);
    if (message) {
      errors[key] = message;
    }
  }
}
